sdv/cag/programmable_constraint.py

```diff
@@ -0,0 +1,196 @@
+                The input metadata.
+
+        Returns:
+            sdv.Metadata:
+                The metadata for the transformed data.
+        """
+        raise NotImplementedError()
+
+    def reverse_transform(self, transformed_data):
+        """Given the transformed data, change it back into the original format.
+
+        Args:
+            transformed_data (dict[pd.DataFrame]):
+                The data to reverse transform.
+
+        Returns:
+            dict[pd.DataFrame]:
+                The reverse transformed data.
+        """
+        raise NotImplementedError()
+
+    def is_valid(self, synthetic_data):
+        """Given synthetic data, validate that it matches the constraint.
+
+        Validate that the synthetic data is valid for the constraint by looking at
+        the parameters as well as any learned information
+
+        Args:
+            synthetic_data (dict[pd.DataFrame]):
+                The data to validate.
+
+        Returns:
+            dict[pd.Series]:
+                For each table, a boolean Series indicating whether or not
+                the corresponding row for the table is valid for this constraint.
+        """
+        raise NotImplementedError()
+
+    def fix_data(self, synthetic_data):
+        """Given synthetic data, fix it to match this constraint.
+
+        Required for use with the DayZSynthesizer. Fixes the synthetic data so that
+        it matches the constraint. You may modify the data or remove invalid data.
+        This method should not require the constraint to be fit first.
+
+        Args:
+            synthetic_data (dict[pd.DataFrame]):
+                The DayZ data to fix.
+
+        Returns:
+            dict[pd.DataFrame]:
+                The “fixed” DayZ data so that it matches the constraint
+        """
+        return synthetic_data
+
+
+class SingleTableProgrammableConstraint(ProgrammableConstraint):
+    """Single table variant of the base programmable constraint class."""
+
+    _is_single_table = True
+
+
+class ProgrammableConstraintHarness(BasePattern):
+    """Constraint interface class for programmable constraints."""
+
+    def __init__(self, programmable_constraint):
+        super().__init__()
+        self.programmable_constraint = programmable_constraint
+        self.table_name = None
+        self._is_single_table = self.programmable_constraint._is_single_table
+
+    def _validate_pattern_with_metadata(self, metadata):
+        if self.programmable_constraint._is_single_table and len(metadata.tables) != 1:
+            raise PatternNotMetError(
+                'SingleTableProgrammableConstraint cannot be used with multi-table metadata. '
+                'Please use the ProgrammableConstraint base class instead.'
+            )
+
+        self.programmable_constraint.validate(metadata)
+
+    def _validate_pattern_with_data(self, data, metadata):
+        if self._is_single_table:
+            data = data[self._get_single_table_name(metadata)]
```

Comment:
why is `self._table_name` not used here?

Reply:
It's not guaranteed to have been set here. We only set `self._table_name` during `fit`, so we have to use the metadata here. This only ever gets called after validating with metadata though, so we will have already caught if the user is trying to pass multiple tables to the single table programmable constraint.

Reply:
but it's available in the _fit below? Is it called by the base fit before _fit is called?